tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ReceiveMessageAsyncIntegration.cs

```diff
@@ -70,7 +76,35 @@ public class ReceiveMessageAsyncIntegration
         /// <param name="state">Calltarget state value</param>
         /// <returns>A response value, in an async scenario will be T of Task of T</returns>
         internal static TResponse OnAsyncMethodEnd<TTarget, TResponse>(TTarget instance, TResponse response, Exception exception, in CallTargetState state)
+            where TResponse : IReceiveMessageResponse
         {
+            if (response.Instance != null && response.Messages != null && response.Messages.Count > 0 && state.State != null)
+            {
+                var dataStreamsManager = Tracer.Instance.TracerManager.DataStreamsManager;
+                if (dataStreamsManager != null && dataStreamsManager.IsEnabled)
+                {
+                    var edgeTags = new[] { "direction:in", $"topic:{(string)state.State}", "type:sqs" };
+                    foreach (var o in response.Messages)
+                    {
+                        var message = o.DuckCast<IMessage>();
+                        if (message == null)
+                        {
+                            continue; // should not happen
+                        }
+
+                        var sentTime = 0;
+                        if (message.Attributes != null && message.Attributes.TryGetValue("SentTimestamp", out var sentTimeStr) && sentTimeStr != null)
+                        {
+                            int.TryParse(sentTimeStr, out sentTime);
+                        }
+
+                        var adapter = new ContextPropagation.MessageAttributesAdapter(message.MessageAttributes);
+                        var parentPathway = dataStreamsManager.ExtractPathwayContext(adapter);
+                        state.Scope.Span.SetDataStreamsCheckpoint(dataStreamsManager, CheckpointKind.Consume, edgeTags, payloadSizeBytes: 0, sentTime, parentPathway);
+                    }
+                }
+            }
+
```

Comment:
Nit: If I'm looking at it right the implementation for this and the sync one are identical, maybe we could extract it to a common method if you think it may be helpful for clarity

Reply:
Yes, I did that in #5120, didn't want to make the diff larger for this PR